scripts/get-all-images.py

```diff
@@ -0,0 +1,223 @@
+
+    Args:
+        repo_name(str): name of the repo to clone
+        branch(str): branch to checkout to, once cloned the repo
+
+    Returns:
+        The path of the downloaded repository
+    """
+    repo_url = f"https://github.com/canonical/{repo_name}.git"
+
+    logging.info(f"Cloning repo {repo_url}")
+    repo = git.Repo.clone_from(repo_url, str(repo_name))
+
+    logging.info(f"Checking out to branch {branch}")
+    repo.git.checkout(branch)
+
+    return repo.working_dir
+
+
+def get_analytics_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a charm developed by us, by:
+    1. Cloning the repo of the charm
+    2. Running the repo's tools/get-images.sh
+    3. Delete the repo
+
+    If the tools/get-images.sh of a repo fails for any reason then this
+    script will also fail.
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_REPO_KEY], app[GH_BRANCH_KEY])
+
+    logging.info(f"Executing repo's {GET_IMAGES_SH} script")
+    try:
+        process = subprocess.run(["bash", "tools/get-images.sh"],
+                                 cwd=repo_dir, capture_output=True,
+                                 text=True, check=True)
+    except subprocess.CalledProcessError as exc:
+        logging.error("Script '%s' for charm '%s' failed: %s",
+                      GET_IMAGES_SH, app["charm"], exc.stderr)
+        raise exc
+    finally:
+        # cleanup
+        shutil.rmtree(repo_dir)
+
+    images = process.stdout.split("\n")
+
+    return images
+
+
+def get_dependency_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a dependency charm by:
+    1. Cloning the repo of the charm
+    2. Looking at its metadata.yaml for "upstream-source" keys
+    3. Delete the repo
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_DEPENDENCY_REPO_KEY],
+                          app[GH_DEPENDENCY_BRANCH_KEY])
+    with open("%s/metadata.yaml" % repo_dir, 'r') as metadata_file:
+        metadata_dict = yaml.safe_load(metadata_file)
+
+    for _, rsrc in metadata_dict["resources"].items():
+        if "type" in rsrc and rsrc["type"] == "oci-image":
+            img = rsrc["upstream-source"]
+            logging.info("Found image %s" % img)
+            images.append(img)
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def cleanup_images(images: list[str]) -> list[str]:
+    """
+    Given a list of OCI registry images ensure
+    1. there are no duplicates
+    2. there are no images with empty name
+    3. the list is sorted
+
+    Args:
+        images: List of images to be processed
+
+    Returns:
+        A new list without duplicate values, empty values and is sorted
+    """
+    images_set = set(images)
+    if '' in images_set:
+        images_set.remove('')
+
+    unique_images = list(images_set)
+    unique_images.sort()
+
+    return unique_images
+
+
+def get_bundle_images(bundle_path: str) -> list[str]:
```

Comment:
should we add here some logging to show the images gathered for each charm? this would help in debugging

Reply:
Nice suggestion! Pushed a commit for this
866d0ca